File: src/openbrain/intent.py
"""Simple intent detection for natural language input.

Classifies a message into an OpenBrain action without requiring an LLM.
Covers the common patterns; anything unrecognised falls back to capture.

# TODO(tailscale): If exposing over the network, add rate limiting here.
# TODO(llm): Replace regex intent matching with a local LLM classifier
#             (e.g. Ollama + llama3) for richer natural language understanding.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum


class Intent(Enum):
    CAPTURE = "capture"
    SEARCH = "search"
    REVIEW = "review"
    STATS = "stats"
    HELP = "help"


@dataclass
class ParsedIntent:
    intent: Intent
    text: str                        # cleaned content/query
    thought_type: str = "note"
    tags: list[str] | None = None


_SEARCH_PATTERNS = re.compile(
    r"^(search|find|look up|what do i know about|recall|remember|remind me|"
    r"have i (thought|noted|written) about|show me|retrieve|query)[:\s]+",
    re.IGNORECASE,
)

_CAPTURE_PATTERNS = re.compile(
    r"^(remember|save|capture|note|log|store|record|add|write down|"
    r"decided?|insight:|learning:|realised?|met |meeting with)[:\s]*",
    re.IGNORECASE,
)

_DECISION_HINT = re.compile(
    r"\b(decided?|chose|choice|picked|going with|will use|won't use|rejected)\b",
    re.IGNORECASE,
)

_INSIGHT_HINT = re.compile(
    r"\b(realised?|learned?|noticed|insight|pattern|key (takeaway|learning))\b",
    re.IGNORECASE,
)

_PERSON_HINT = re.compile(
    r"\b(met |talked? (to|with)|called?|spoke (to|with)|email(ed)?)\b",
    re.IGNORECASE,
)

_MEETING_HINT = re.compile(
    r"\b(meeting|standup|call|sync|retrospective|1:1|one.on.one)\b",
    re.IGNORECASE,
)

_REVIEW_PATTERNS = re.compile(
    r"^(weekly review|week review|review( the)? week|what happened this week"
    r"|this week|past week|last 7 days|summarise( the)? week)",
    re.IGNORECASE,
)

_STATS_PATTERNS = re.compile(
    r"^(stats|statistics|how many (thoughts|memories|notes)|"
    r"brain stats|knowledge base stats|count)",
    re.IGNORECASE,
)

_HELP_PATTERNS = re.compile(
    r"^(help|commands|what can you do|\?+|how (do|does) (this|it) work)",
    re.IGNORECASE,
)


def _infer_type(text: str) -> str:
    if _DECISION_HINT.search(text):
        return "decision"
    if _INSIGHT_HINT.search(text):
        return "insight"
    if _PERSON_HINT.search(text):
        return "person"
    if _MEETING_HINT.search(text):
        return "meeting"
    return "note"


def parse(message: str) -> ParsedIntent:
    """Parse a natural language message into a structured intent."""
    msg = message.strip()

    if _HELP_PATTERNS.match(msg):
        return ParsedIntent(intent=Intent.HELP, text=msg)

    if _STATS_PATTERNS.match(msg):
        return ParsedIntent(intent=Intent.STATS, text=msg)

    if _REVIEW_PATTERNS.match(msg):
        return ParsedIntent(intent=Intent.REVIEW, text=msg)

    search_match = _SEARCH_PATTERNS.match(msg)
    if search_match:
        query = msg[search_match.end():].strip()
        return ParsedIntent(intent=Intent.SEARCH, text=query)

    capture_match = _CAPTURE_PATTERNS.match(msg)
    if capture_match:
        content = msg[capture_match.end():].strip() or msg
        return ParsedIntent(
            intent=Intent.CAPTURE,
            text=content,
            thought_type=_infer_type(content),
        )

    # Default: try to capture anything that looks like a statement
    # Short questions fall back to search
    if msg.endswith("?") or msg.lower().startswith(("what", "who", "when", "where", "how", "why")):
        return ParsedIntent(intent=Intent.SEARCH, text=msg.rstrip("?"))

    return ParsedIntent(
        intent=Intent.CAPTURE,
        text=msg,
        thought_type=_infer_type(msg),
    )


HELP_TEXT = """
**OpenBrain** — your personal knowledge base

**Capture a thought:**
> decided to use Redis for session caching
> realised that deploys on Fridays are always risky
> met Sarah Chen, she runs engineering at Acme
> remember: the API rate limit is 1000 req/min

**Search:**
> search: Redis decisions
> what do I know about Sarah?
> find: deployment lessons

**Weekly review:**
> weekly review
> what happened this week?

**Stats:**
> stats
> how many thoughts?

Anything that looks like a statement gets captured automatically.
Anything that looks like a question triggers a search.
""".strip()

File: internal/intent/intent.go
// Package intent classifies natural language messages into OpenBrain actions
// using regex pattern matching — no LLM required.
package intent

import (
	"regexp"
	"strings"
)

// Intent represents a classified user action.
type Intent string

const (
	Capture   Intent = "capture"
	Search    Intent = "search"
	Review    Intent = "review"
	Stats     Intent = "stats"
	Help      Intent = "help"
	Supersede Intent = "supersede"
	Extract   Intent = "extract"
	Reload    Intent = "reload"
)

// ParsedIntent holds the classified intent and extracted content.
type ParsedIntent struct {
	Intent         Intent
	Text           string
	ThoughtType    string
	Tags           []string
	SupersedeQuery *string
}

const DeepCaptureThreshold = 200

var (
	searchPatterns = regexp.MustCompile(
		`(?i)^(search|find|look up|what do i know about|recall|remember|remind me|` +
			`have i (thought|noted|written) about|show me|retrieve|query)[:\s]+`)

	capturePatterns = regexp.MustCompile(
		`(?i)^(remember|save|capture|note|log|store|record|add|write down|` +
			`decided?|insight:|learning:|realised?|met |meeting with)[:\s]*`)

	decisionHint = regexp.MustCompile(
		`(?i)\b(decided?|chose|choice|picked|going with|will use|won't use|rejected)\b`)

	insightHint = regexp.MustCompile(
		`(?i)\b(realised?|learned?|noticed|insight|pattern|key (takeaway|learning))\b`)

	personHint = regexp.MustCompile(
		`(?i)\b(met |talked? (to|with)|called?|spoke (to|with)|email(ed)?)\b`)

	meetingHint = regexp.MustCompile(
		`(?i)\b(meeting|standup|call|sync|retrospective|1:1|one.on.one)\b`)

	reviewPatterns = regexp.MustCompile(
		`(?i)(weekly review|week review|review( the)? week|what happened this week` +
			`|this week|past week|last 7 days|summarise( the)? week` +
			`|give me a (weekly |week )?review|show me (the )?(weekly |week )?review)`)

	statsPatterns = regexp.MustCompile(
		`(?i)(^stats?$|statistics|how many (thoughts|memories|notes)|` +
			`brain stats|knowledge base stats|^count$` +
			`|send me (some |the |my )?(stats?|status)|give me (some |the |my )?(stats?|status)` +
			`|show me (some |the |my )?(stats?|status)|what (are|is) (my |the )?(stats?|status))`)

	supersedePatterns = regexp.MustCompile(
		`(?i)^(actually[,:]?\s*|update:\s*|correction:\s*|changed?:\s*|no longer[,:]?\s*|now instead[,:]?\s*)`)

	extractPrefix = regexp.MustCompile(`(?i)^extract:\s*`)

	helpPatterns = regexp.MustCompile(
		`(?i)(^help$|^commands$|^what can you do|^\?+$|^how (do|does) (this|it) work)`)
)

// InferType guesses the thought type from content text.
func InferType(text string) string {
	if decisionHint.MatchString(text) {
		return "decision"
	}
	if insightHint.MatchString(text) {
		return "insight"
	}
	if personHint.MatchString(text) {
		return "person"
	}
	if meetingHint.MatchString(text) {
		return "meeting"
	}
	return "note"
}

// Parse classifies a natural language message into a structured intent.
func Parse(message string) ParsedIntent {
	msg := strings.TrimSpace(message)

	if helpPatterns.MatchString(msg) {
		return ParsedIntent{Intent: Help, Text: msg, ThoughtType: "note"}
	}

	if statsPatterns.MatchString(msg) {
		return ParsedIntent{Intent: Stats, Text: msg, ThoughtType: "note"}
	}

	if reviewPatterns.MatchString(msg) {
		return ParsedIntent{Intent: Review, Text: msg, ThoughtType: "note"}
	}

	// Extract prefix: explicit deep capture
	if loc := extractPrefix.FindStringIndex(msg); loc != nil {
		content := strings.TrimSpace(msg[loc[1]:])
		return ParsedIntent{Intent: Extract, Text: content, ThoughtType: "note"}
	}

	// Supersede: "actually,...", "update:...", "correction:..."
	if loc := supersedePatterns.FindStringIndex(msg); loc != nil {
		content := strings.TrimSpace(msg[loc[1]:])
		q := content
		return ParsedIntent{
			Intent:         Supersede,
			Text:           content,
			ThoughtType:    InferType(content),
			SupersedeQuery: &q,
		}
	}

	if loc := searchPatterns.FindStringIndex(msg); loc != nil {
		query := strings.TrimSpace(msg[loc[1]:])
		return ParsedIntent{Intent: Search, Text: query, ThoughtType: "note"}
	}

	if loc := capturePatterns.FindStringIndex(msg); loc != nil {
		content := strings.TrimSpace(msg[loc[1]:])
		if content == "" {
			content = msg
		}
		return ParsedIntent{Intent: Capture, Text: content, ThoughtType: InferType(content)}
	}

	// Default: questions -> search, long text -> extract, else capture
	lower := strings.ToLower(msg)
	if strings.HasSuffix(msg, "?") ||
		strings.HasPrefix(lower, "what") ||
		strings.HasPrefix(lower, "who") ||
		strings.HasPrefix(lower, "when") ||
		strings.HasPrefix(lower, "where") ||
		strings.HasPrefix(lower, "how") ||
		strings.HasPrefix(lower, "why") {
		return ParsedIntent{Intent: Search, Text: strings.TrimRight(msg, "?"), ThoughtType: "note"}
	}

	if len(msg) > DeepCaptureThreshold {
		return ParsedIntent{Intent: Extract, Text: msg, ThoughtType: "note"}
	}

	return ParsedIntent{Intent: Capture, Text: msg, ThoughtType: InferType(msg)}
}
